src/devices/SmartShunt.cpp

```diff
@@ -4,326 +4,414 @@
 // https://opensource.org/licenses/MIT
 
 #include "devices/SmartShunt.h"
+#include "devices/DeviceUtils.hpp"
+#include "BscSerial.h"
 #include "BmsData.h"
 #include "mqtt_t.h"
 #include "log.h"
 #include "WebSettings.h"
+#include "Utility.h"
 
 static const char *TAG = "SMARTSHUNT";
 
 static Stream *mPort;
 static uint8_t u8_mDevNr;
-static uint8_t mDataMappingNr;
-
-enum States {
-  IDLE,
-  RECORD_BEGIN,
-  RECORD_NAME,
-  RECORD_VALUE,
-  CHECKSUM,
-  RECORD_HEX
-};
-
-int mState;                                 // current state
-char * mTextPointer;                        // pointer to the private buffer we're writing to, name or value
-char mName[9];                              // buffer for the field name
-char mValue[33];                            // buffer for the field value
-
-static constexpr char checksumTagName[] = "CHECKSUM";
 
-uint8_t u8_tSoc;   // %
-int16_t i16_tVolt; // mV
-int16_t i16_tCurr; // mA
-
-uint8_t rxValues=0;
-
-// Test
-static uint16_t errCntSmartShunt=0;
+static void      getDataFromBms(deviceUtils::DeviceUtils &devUtils, BscSerial *bscSerial, uint16_t ID_Get);
+static bool      recvAnswer(deviceUtils::DeviceUtils &devUtils, uint8_t * t_outMessage, uint16_t ID_Get);
+static void      parseMessage(deviceUtils::DeviceUtils &devUtils, uint8_t * t_message);
+static bool      hexIsValid(deviceUtils::DeviceUtils &devUtils, const uint8_t* buffer, int size);
+static bool      getBMSTelegramm(deviceUtils::DeviceUtils &devUtils, BscSerial *bscSerial, uint16_t ID_Get);
 
+enum SM_readData {SEARCH_START, SEARCH_END};
 
 static serialDevData_s *mDevData;
 
-//https://www.victronenergy.com/live/vedirect_protocol:faq
+static uint8_t u8_DataPointer;
+static uint8_t mDataMappingNr;
 
+//https://www.victronenergy.com/live/vedirect_protocol:faq
 
 
-void newLabelRecv(char * mName, char * mValue)
+bool SmartShunt_readBmsData(BscSerial *bscSerial, Stream *port, uint8_t devNr, serialDevData_s *devData)
 {
-  int32_t val;
-  //BSC_LOGI(TAG,"New value: name=%s, val=%s",mName,mValue);
-
-  if(strcmp_P(mName, "SOC") == 0)
-  {
-    sscanf(mValue, "%ld", &val);
-    u8_tSoc=(uint8_t)(val/10);
-    rxValues|=RX_VAL_SOC;
-    //BSC_LOGI(TAG,"SoC=%i, rxValues=%i",u8_tSoc,rxValues);
-  }
-  else if(strcmp_P(mName, "V") == 0)
-  {
-    sscanf(mValue, "%ld", &val); // mV
-    i16_tVolt=val/10;
-    rxValues|=RX_VAL_U;
+  bool ret = true;
+  mDevData = devData;
+  mPort = port;
+  u8_mDevNr = devNr;
+  uint8_t response[SMARTSHUNT_MAX_ANSWER_LEN];
```

Comment:
Du hast hier ein response Array und in getBMSTelegramm() deklarierst du nich einmal ein response Array. 
Dieses hier ist eigentlich überflüssig.